internal/worker/server.go

```diff
@@ -131,11 +133,68 @@ func (s *Server) JobStatus(id uuid.UUID, result interface{}) (*JobStatus, []uuid
 		}
 	}
 
-	// For backwards compatibility: OSBuildJobResult didn't use to have a
-	// top-level `Success` flag. Override it here by looking into the job.
+	if r, ok := result.(*KojiInitJobResult); ok {
+		if r.JobResult.JobError == nil && r.KojiError != "" {
+			r.JobResult = JobResult{
+				JobError: clienterrors.WorkerClientError(clienterrors.ErrorOldResultCompatible, r.KojiError),
+			}
+		}
+	}
+
+	if r, ok := result.(*KojiFinalizeJobResult); ok {
+		if r.JobResult.JobError == nil && r.KojiError != "" {
+			r.JobResult = JobResult{
+				JobError: clienterrors.WorkerClientError(clienterrors.ErrorOldResultCompatible, r.KojiError),
+			}
+		}
+	}
+
+	if r, ok := result.(*OSBuildKojiJobResult); ok {
+		if r.JobResult.JobError == nil && r.KojiError != "" {
+			r.JobResult = JobResult{
+				JobError: clienterrors.WorkerClientError(clienterrors.ErrorOldResultCompatible, r.KojiError),
+			}
+		}
+	}
+
+	if r, ok := result.(*ManifestJobByIDResult); ok {
+		if r.JobResult.JobError == nil && r.Error != "" {
+			r.JobResult = JobResult{
+				JobError: clienterrors.WorkerClientError(clienterrors.ErrorOldResultCompatible, r.Error),
+			}
+		}
+	}
+
+	if r, ok := result.(*DepsolveJobResult); ok {
+		if r.JobResult.JobError == nil && r.Error != "" {
+			var jobErr *clienterrors.Error
+			if r.ErrorType == DepsolveErrorType {
+				jobErr = clienterrors.WorkerClientError(clienterrors.ErrorDNFError, r.Error)
```

Comment:
Based on the if statement, shouldn't this branch use the `clienterrors.ErrorDepsolveJob` instead of the else branch?

Reply:
@thozza, do you have a suggestion for the code in the `else` branch here?
I think the same thing would apply for this bit of code too:
https://github.com/osbuild/osbuild-composer/pull/2192/files#diff-cdbd73affc81d7f2669bf94f8b6fecae6502db28431663452e9fac613728bff1L488-R509

Reply:
And I guess here too:
https://github.com/osbuild/osbuild-composer/pull/2192/files#diff-e0ac8df111d90787f24679627e7beea4774877480640e1494d296a4b2096b1b4R44-R49

Reply:
I mean the `clienterrors.ErrorDNFError` in the `else` branch looked reasonable to me. 

To your other questions, I guess you should distinguish if the actual depsolving of the package set failed, or if there is some other DNF error and return the relevant error type. Using `clienterrors.ErrorDNFError` for non-depsolving errors even in the depsolving job seems OK to me.

Reply:
I seemed to have caused a bunch of confusion with https://github.com/osbuild/osbuild-composer/pull/2192#discussion_r786743109

So how about https://github.com/Gundersanne/osbuild-composer/commit/b6bb949bfcbb502d41b9a969d9b37f249152ecab.

rpmmd classifies anything that came from dnf-json (where the http call didn't succeed and returned normally) as a "DNFError". Crrently anything classified as a DNFError is considered bad user input (though I'm not entirely convinced this should be the case).

If dnf-json crashes I'm not actually sure where it'll fail, will httpc.Post fail? Regardless, if it's not a "DNFError" it's an RPMMDError. And then let's drop the confusing ErrorDepsolveJob.

I think user input failures are definitely ErrorDNFDepsolveError, corresponding with https://dnf.readthedocs.io/en/latest/api_exceptions.html#dnf.exceptions.DepsolveError ; unsure about the MarkingErrors.

Reply:
We could just not change the behaviour in this PR and keep the distinction DNFError vs RPMMDError only, and then if it's a DNFError we regard it as user input failure always like we do now. And ameliorate the behavior in some future work.